core/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -958,16 +977,17 @@ private void ensureWriteAllowed(Engine.Operation op) throws IllegalIndexShardSta
         IndexShardState state = this.state; // one time volatile read
 
         if (origin == Engine.Operation.Origin.PRIMARY) {
-            // for primaries, we only allow to write when actually started (so the cluster has decided we started)
-            // otherwise, we need to retry, we also want to still allow to index if we are relocated in case it fails
-            if (state != IndexShardState.STARTED && state != IndexShardState.RELOCATED) {
+            if (writeAllowedStatesForPrimary.contains(state) == false) {
                 throw new IllegalIndexShardStateException(shardId, state, "operation only allowed when started/recovering, origin [" + origin + "]");
```

Comment:
we can just use the enumSet.toString here (assuming it gives nice output). This will make sure we are never out of sync and I think it's OK to have a technical message here.